Clarify one sided payment rules (#2837)
tari-project/tari: RFC/src/RFC-0201_TariScript.md

Alice owns \\( C_a \\) and provides the required script to spend the UTXO as was described in the previous cases.

Alice needs a public key from Bob, \\( K_{Sb} \\) to complete the one-sided transaction. This key can be obtained
out-of-band, and might typically be Bob's wallet public key on the Tari network.

Bob requires the value \\( v_b \\) and blinding factor \\( k_b \\) to claim his payment, but he needs to be able to claim it without asking Alice for them.

> This information can be obtained by using diffie-hellman and bulletproof rewinding. If the blinding factor \\( k_b \\) was calculated with diffie-hellman using the offset public keypair, (\\( k_{Ob} \\),\\( K_{Ob} \\)) as sender keypair and

Reviewer: ```suggestion
This information can be obtained by using Diffie-Hellman and bulletproof rewinding. If the blinding factor \\( k_b \\) was calculated with diffie-hellman using the offset public keypair, (\\( k_{Ob} \\),\\( K_{Ob} \\)) as sender keypair and
```

the keypair, (\\( k_{Sb} \\),\\( K_{Sb} \\)) as the receiver keypair, the blinding factor \\( k_b \\) can be securely calculated without communication.

Alice uses Bob's public key to create a shared secret, \\( k_b \\) for the output commitment, \\( C_b \\), using
Diffie-Hellman key exchange.

Alice calculates \\( k_b \\) as
$$
    k_b = k_{Ob} * {K_Sb}
$$

> Next Alice next uses Bulletproof rewinding to encrypt the value \\( v_b \\) into the the Bulletproof for the commitment \\( C_b \\). For this she uses (\\( k_{rewind} =  Hash(k_{b}) \\) as the encryption key.

Reviewer 2: We need two private keys here: `pvt_rewind_key` and `pvt_blinding_key`. This could be something like:
```
k_{rewind} =  Hash(k_{b})
k_{blinding} =  Hash(Hash(k_{b}) | k_{b}))
```
Please see the Bulletproofs code snippet for the full explanation.

> Next alice next uses bulletproof rewinding to encrypt the value \\( v_b \\) into the the bulletproof for the commitment \\( C_b \\). For this she uses (\\( k_{rewind} =  Hash(k_{Ob}) \\) as the encryption key.

Reviewer: ```suggestion
Next Alice next uses bulletproof rewinding to encrypt the value \\( v_b \\) into the the bulletproof for the commitment \\( C_b \\). For this she uses (\\( k_{rewind} =  Hash(k_{Ob}) \\) as the encryption key.
```

In this case, the script hash is analogous to an address in Bitcoin or Monero. Bob's wallet can scan the blockchain
looking for hashes that he would know how to resolve. For all outputs that he discovers this way, Bob would need to know
who the sender is so that he can derive the shared secret.  

This information can be obtained without needing to communicate with the sender. If the blinding factor \\( k_b \\) was calculated with diffie-hellman using the offset public keypair, (\\( k_{Ob} \\),\\( K_{Ob} \\)) as sender keypair and
the keypair, (\\( k_{Sb} \\),\\( K_{Sb} \\)) as the receiver keypair, the blinding factor \\( k_b \\) can be securely calculated without communication between the 2 parties.

Alice calculates \\( k_b \\) as
$$
    k_b = k_{Ob} * {K_Sb}
$$

Bob calculates \\( k_b \\) as
$$
    k_b = K_{Ob} * {k_Sb}
$$

For Bob's part, when he discovers one-sided payments to himself, he should spend them to new outputs using a traditional
transaction to thwart any potential horizon attacks in the future.

> One other problem with one sided communication is the communication of the value of the transaction being sent, and the wallet recovery of one sided payments. We can solve both of these by using (\\( k_{Ob} \\).

Reviewer: ```suggestion
One other problem with one sided communication is the communication of the value of the transaction being sent, and the wallet recovery of one sided payments. We can solve both of these by using (\\( k_b \\).
```

Author: We should be using k_{Ob}, the offset pubkey. k_{b} is the blinding factor.

Reviewer: Bob does not know `k_{Ob}` though?

Author: Yes, He only knows `K_{Ob}` and `k_S`

Reviewer: The wording is ambiguous, but the next line is correct, so will let it pass

In this case, the script hash is analogous to an address in Bitcoin or Monero. Bob's wallet can scan the blockchain
looking for hashes that he would know how to resolve. For all outputs that he discovers this way, Bob would need to know
who the sender is so that he can derive the shared secret.  

This information can be obtained without needing to communicate with the sender. If the blinding factor \\( k_b \\) was calculated with diffie-hellman using the offset public keypair, (\\( k_{Ob} \\),\\( K_{Ob} \\)) as sender keypair and

> the keypair, (\\( k_{Sb} \\),\\( K_{Sb} \\)) as the receiver keypair, the blinding factor \\( k_b \\) can be securely calculated without communication between the 2 parties.

Reviewer 3: Not super necessary, but might be clearer to be 100% specific. 
i.e.
The sender calculates k_b = k_Ob * K_Sb, and the receiver calculates k_b = K_Ob * k_Sb

Reviewer 2: Agreed

Reviewer: How does the receiver know the value though?

Reviewer 2: The value must be sent out of band.

Reviewer: if the value is sent out of band, why not send the whole commitment out of band? Isn't the purpose here to not do out of band? 
"This information can be obtained with need to communicate with the sender"

    \so = k_{Sa} - k_{Ob} \cdot \HU_b
$$

For the script hash, she provides the hash of a script that locks the output to Bob's public key, `PushPubkey(K_Sb)`.
This script will only resolve successfully if the spender can provide a valid signature as input that demonstrates proof
of knowledge of \\( k_{Sb} \\) which only Bob knows.

Any base node can now verify that the transaction is complete, verify the signature on the script, and verify the script
offset.

For Bob to claim his commitment he will scan the blockchain for a known script hash because he knowns that the script will be `PushPubkey(K_Sb)` he can scan for that hash. In this case, the script hash is analogous to an address in Bitcoin or Monero. Bob's wallet can scan the blockchain
looking for hashes that he would know how to resolve. For all outputs that he discovers this way, Bob would need to know
who the sender is so that he can derive the shared secret.

When Bob's wallet spots a known hash he requires he requires the blinding factor, \\( k_b \\) and the value \\( v_b \\). First he uses Diffie-Hellman to calculate \\( k_b \\). 

Bob calculates \\( k_b \\) as
$$
    k_b = K_{Ob} * {k_Sb}
$$

> Next Bob's wallet calculates \\( k_{rewind} \\), using \\( k_{rewind} =  Hash(k_{b})\\) and uses that to rewind the Bulletproof to get the value \\( v_b \\). 

Reviewer 2: **Edit:** _Strike that_ ~~We equate the private rewind key to the hash of the blinding factor. Bulletproof rewinding is done with the public rewind key, not the private rewind key.~~

Reviewer 2: We actually need to convey two values, not only one. To be clear on the previous comment, Bulletproof rewinding is done with the two public rewind keys, not the private rewind keys.

From [Tari fork of Bulletproofs]( https://github.com/tari-project/bulletproofs/blob/a18a3d2c91cea4e5f999446a566a4b0af7c9e187/tests/range_proof.rs#L154)
``` rust
    // The two rewind keys can be shared with a trusted 3rd party, whom will be able to extract
    // the value of the commitment as well as the extra data
    let pub_rewind_key_1 =
        RistrettoPoint::from(&pvt_rewind_key * &RISTRETTO_BASEPOINT_TABLE).compress();
    let pub_rewind_key_2 =
        RistrettoPoint::from(&pvt_blinding_key * &RISTRETTO_BASEPOINT_TABLE).compress();
    // The rewind nonces are derived from the public rewind keys and the Pedersen commitment
    let rewind_nonce_1 = get_rewind_nonce_from_pub_key(&pub_rewind_key_1, &committed_value);
    let rewind_nonce_2 = get_rewind_nonce_from_pub_key(&pub_rewind_key_2, &committed_value);
    // Get secret nonces from the private keys - this emulates the owner of a wallet
    let blinding_nonce_1 = get_secret_nonce_from_pvt_key(&pvt_rewind_key, &committed_value);
    let blinding_nonce_2 = get_secret_nonce_from_pvt_key(&pvt_blinding_key, &committed_value);

    // Get Value Test 1 - provide correct rewind nonces
    let mut rewind_transcript = Transcript::new(b"Bulletproof-Rewind Test");
    assert_eq!(
        proof.rewind_single_get_value_only(
            &bp_gens,
            &mut rewind_transcript,
            &committed_value,
            64,
            &rewind_nonce_1,
            &rewind_nonce_2,
        ),
        Ok((confidential_value, proof_message))
    );
```

> Next Bob's wallet calculates \\( k_{rewind} \\), using \\( k_{rewind} =  Hash(k_{Ob})\\) and uses that to rewind the bulletproof to get the value \\( v_b \\). 

Reviewer: Where does Bob find out `k_{0}b` in order to hash it?

For Bob's part, when he discovers one-sided payments to himself, he should spend them to new outputs using a traditional
transaction to thwart any potential horizon attacks in the future.

One other problem with one sided communication is the communication of the value of the transaction being sent, and the wallet recovery of one sided payments. We can solve both of these by using (\\( k_{Ob} \\).

> Because this is a secret key known by Alice and Bob and no one else. We hash this key as a new key, (\\( k_{rewind} =  Hash(k_{Ob}) \\). We know have a rewinding secret key for use in bulletproof rewinding. And now

Reviewer 2: We ~~know~~ _now_ have ...

Reviewer: ```suggestion
Because this is a secret key known by Alice and Bob and no one else. We hash this key as a new key, (\\( k_{rewind} =  Hash(k_b) \\). We know have a rewinding secret key for use in bulletproof rewinding. And now
```

> we can encrypt the value (\\( v \\) inside of the bulletproof rewinding, allowing Bob to unlock his payment without revealing any information publicly.

Reviewer 2: ~~we can encrypt the value (\\( v \\) inside of the bulletproof rewinding~~
we use the standard Bulletproof rewinding scheme to encrypt the value (\\( v \\) inside the Bulletproof range proof,

> we can encrypt the value (\\( v \\) inside of the bulletproof rewinding, allowing Bob to unlock his payment without revealing any information publicly.

Reviewer 2: ```suggestion
we use the standard Bulletproof rewinding scheme to encrypt the value (\( v \) inside the Bulletproof range proof,, allowing Bob to unlock his payment without revealing any information publicly. (See [Bulletproof rewinding](https://github.com/dalek-cryptography/bulletproofs/issues/335) for context)).
```

Reviewer: Should be "We now use the standard...". Don't need to start a sentence with "And"

Author: Do we have a link explaining it in our own repo, not Dalek's?

Reviewer 2: We can always update it later when we have documented it ourselves.

> 

Reviewer 2: I would add something like this:

-----------

In summary, receiving party scanning the blockchain for one-sided payments therefore involves:
- Scan all UTXOs trying to unlock the payment script hashes; if the script resolves when inserting the receiver public key, it is for the receiver.
- With those identified UTXOs, do Bulletproof rewinding with the range of registered sender public rewind keys ((\( k_{rewind}G \) as per above). If the proof message matches, the rewound value is more than likely the correct value of the commitment.
- Test opening of the commitment using the value obtained from rewinding and the spending key from the Diffie-Hellman exchange; if the commitment can be opened, it is intended for the receiver.

Reviewer: Don't agree with point 2. Point 2 should be rewound using the hash of the blinding factor

Reviewer 2: **Edit: ** _Strike that_ ~~We equate the private rewind key to the hash of the blinding factor. Bulletproof rewinding is done with the public rewind key, not the private rewind key.~~

Please see the explanation below.
